Add new fields for kubeproxy and kubelet fields in cluster config

terciodemelo opened the pull request:
This PR aims to solve #29. The fields I want to add are:

```
kubelet:
- fail_swap_on: <bool>

kubeproxy:
  extra_binds:
  - <strings>
```

The `fail_swap_on` works fine, going to the cluster config as expected, but the `kubeproxy.extra_binds` doesn't.

This is the slice of the output of `terraform apply` that shows the fields I want to add:

```
rke_config.0.services.0.kubelet.0.fail_swap_on:                                                 "false"
rke_config.0.services.0.kubeproxy.#:                                                            "1"
rke_config.0.services.0.kubeproxy.0.extra_args.%:                                               "1"
rke_config.0.services.0.kubeproxy.0.extra_args.ipvs-scheduler:                                  "lc"
rke_config.0.services.0.kubeproxy.0.extra_binds.#:                                              "2"
rke_config.0.services.0.kubeproxy.0.extra_binds.0:                                              "/usr/lib/modules:/lib/modules"
rke_config.0.services.0.kubeproxy.0.extra_binds.1:                                              "/var/run/dbus/system_bus_socket:/var/run/dbus/system_bus_socket"
rke_config.0.ssh_agent_auth:                                                                    "false"
```

But the `extra_binds` of `kubeproxy` are not appearing in the cluster config as expected, and no error message is shown as well. I'm going to dig the RKE API to find something related that can help me solve this issue. Any help is welcome

terciodemelo commented:
I've verified that the payload to RKE from Terraform doesn't include `extraBinds` of `kubeproxy`, even though it includes `extraArgs` as shown below. So I'm guessing the code that builds this json payload is ignoring extra binds

```json
...
"kubeproxy": {
    "extraArgs":{"ipvs-scheduler":"lc"},
    "type":"/v3/schemas/kubeproxyService"
}
```

rawmind0 commented:
Hi @terciodemelo, thanks for the PR.

With these code, none of added fields will work correctly. You need to update flatten and expand functions (for kubelet and kubeproxy) to map data correctly from rancher obj to tf info and the reverse.

https://github.com/rancher/terraform-provider-rancher2/blob/master/rancher2/cluster_rke_config_services.go#L235
https://github.com/rancher/terraform-provider-rancher2/blob/master/rancher2/cluster_rke_config_services.go#L256
https://github.com/rancher/terraform-provider-rancher2/blob/master/rancher2/cluster_rke_config_services.go#L394
https://github.com/rancher/terraform-provider-rancher2/blob/master/rancher2/cluster_rke_config_services.go#L416 

Anyway, it would be nice to add all services rke fields to the terraform provider. :)

terciodemelo commented:
Thank you @rawmind0!

I've updated the flatten and expand functions for kubeproxy only, which is my immediate urgent need right now. It would indeed be nice to have more RKE fields to terraform provider, I may come back in the near future with more of them!

I've also added the added fields descriptions to `cluster.html.markdown` for documentation, so for now I'll consider this PR done if it's alright with you guys :)

rawmind0 commented:
I don't see `fail_swap_on` on flatten nor expander functions...

terciodemelo commented:
I thought it's not necessary because it's a bool, and in fact it goes fine to the yaml

terciodemelo commented:
Gonna add it, be right back 🏃

terciodemelo commented:
Added and tested it, sorry for having missed it 🤷‍♂️

rawmind0 commented:
@terciodemelo , finally i submitted a PR adding all fields for rke services.

This PR is superseded by https://github.com/rancher/terraform-provider-rancher2/pull/32
